lib/oplog/tail.go

```diff
@@ -143,56 +172,137 @@ func (tailer *Tailer) tailOnce(out chan<- *redispub.Publication, stop <-chan boo
 
-		for iter.Next(&rawData) {
-			ts, pubs := tailer.unmarshalEntry(rawData)
+		for {
+			gotResult, didTimeout, didLosePosition, err := readNextFromCursor(query)
 
-			if ts != nil {
-				lastTimestamp = *ts
-			}
+			if gotResult {
+				decodeErr := query.Decode(&rawData)
+				if decodeErr != nil {
+					log.Log.Errorw("Error decoding oplog entry", "error", decodeErr)
+
+				}
+
+				ts, pubs := tailer.unmarshalEntry(rawData)
+
+				if ts != nil {
+					lastTimestamp = *ts
+				}
+
+				for _, pub := range pubs {
+					if pub != nil {
+						out <- pub
+					} else {
+						log.Log.Error("Nil Redis publication")
+					}
+				}
+			} else if didTimeout {
+				log.Log.Info("Oplog cursor timed out, will retry")
 
-			for _, pub := range pubs {
-				if pub != nil {
-					out <- pub
-				} else {
-					log.Log.Error("Nil Redis publication")
+				query, queryErr = issueOplogFindQuery(oplogCollection, lastTimestamp)
+
+				if queryErr != nil {
+					log.Log.Errorw("Error issuing tail query", "error", queryErr)
+					return
 				}
+
+				break
+			} else if didLosePosition {
+				// Our cursor expired. Make a new cursor to pick up from where we
+				// left off.
+				query, queryErr = issueOplogFindQuery(oplogCollection, lastTimestamp)
+
+				if queryErr != nil {
+					log.Log.Errorw("Error issuing tail query", "error", queryErr)
+					return
+				}
+
+				break
+			} else if err != nil {
+				log.Log.Errorw("Error from oplog iterator",
+					"error", query.Err())
+
+				closeCursor(query)
+
+				return
+			} else {
+				log.Log.Errorw("Got no data from cursor, but also no error. This is unexpected; restarting query")
+
+				closeCursor(query)
+
+				return
 			}
 		}
+	}
+}
 
-		if iter.Err() != nil {
-			log.Log.Errorw("Error from oplog iterator",
-				"error", iter.Err())
+func readNextFromCursor(cursor *mongo.Cursor) (gotResult bool, didTimeout bool, didLosePosition bool, err error) {
+	ctx, cancel := context.WithTimeout(context.Background(), config.MongoQueryTimeout())
+	defer cancel()
 
-			closeErr := iter.Close()
-			if closeErr != nil {
-				log.Log.Errorw("Error from closing oplog iterator",
-					"error", closeErr)
-			}
+	gotResult = cursor.Next(ctx)
+	err = cursor.Err()
 
-			return
+	if err != nil {
+		// Wait briefly before determining whether the failure was a timeout: because
+		// the MongoDB go driver passes the context on to lower-level networking
+		// components, it's possible for the query to fail *before* the context
+		// is marked as timed-out
+		time.Sleep(100 * time.Millisecond)
+		didTimeout = ctx.Err() != nil
+
+		// check if the error is a position-lost error. These errors are best handled
+		// by just re-issueing the query on the same connection; no need to surface
+		// a major error and re-connect to mongo
+		// From: https://github.com/rwynn/gtm/blob/e02a1f9c1b79eb5f14ed26c86a23b920589d84c9/gtm.go#L547
+		var serverErr mongo.ServerError
+		if errors.As(err, &serverErr) {
+			// 136  : cursor capped position lost
```

Comment:
OK... this seems version-specific...do these error codes exist since, say, mongodb 3.6? 4.0? And if 5.0 adds  more "lost position" error codes, we could miss that. But I guess in that case, we would just raise a generic error, and not optimize the lost position error.

Reply:
I agree. This is just an optimization for errors we can handle by just re-creating the cursor.